### ISSUE TITLE
HTTPS SNI connections

### ISSUE THREAD

**Issue by pvogt-svc** [mitre/HTTP-Proxy-Servlet#61]
I can not connect to an HTTPS server with SNI with version 1.6 of the HTTP-Proxy-Servlet.

I locally patched the `org.mitre.dsmiley.httpproxy.ProxyServlet.createHttpClient(HttpParams)` method to enable SNI like the page [1] describes.

I used the following code:

    SSLContext sslcontext = SSLContexts.createSystemDefault();
    SSLSocketFactory sslsf = new SSLSocketFactory(sslcontext) {
      // here the code from [1]
    }
    HttpClientBuilder builder = HttpClients.custom().setConnectionManager(new PoolingHttpClientConnectionManager()).setSSLSocketFactory(sslsf);
    builder.disableCookieManagement();
    if (hcParams.isParameterFalse(ClientPNames.HANDLE_REDIRECTS)) {
      builder.disableRedirectHandling();
    }
    return builder.build();

I'm not sure if this is compatible with the code for `org.apache.http.impl.client.SystemDefaultHttpClient` in `createHttpClient()`.

[1] https://wiki.apache.org/HttpComponents/SNISupport

**Comment by dsmiley** [comment]
Thanks for sharing this.  I'm not sure what you mean by if this is compatible with org.apache.http.impl.client.SystemDefaultHttpClient; -- do you mean equivalent?  It would appear that way based on what the javadocs say and a little poking around I did.

**Comment by pvogt-svc** [comment]
Yes, I just tried to get it working and didn't really check what exactly `SystemDefaultHttpClient` does and what kind of features you need from it. I also just tried to get the basic properties of `hcParams` set by the `HttpClientBuilder` to get the resulting code past the JUnit tests. 

I also didn't search too hard for an easier way to get this working. The code I used is relying on `commons-beanutils` which may be a dependency that you don't want. At least Oracle is giving other example code (https://docs.oracle.com/javase/8/docs/technotes/guides/security/jsse/JSSERefGuide.html#SNIExtension) but I didn't try to integrate it with the apache httpclient. So there may be better / easier solutions.

On our system the code worked with JDK 7 and an apache server that used SNI. That SNI configuration was more by mistake than intention, so we reversed the configuration on the apache server. But I used the time to see if our little proxy application (used in the development system) that uses your servlet works. That resulted in this issue. We may use SNI in the future therefore I created this issue with the code I used.

Personal note: your proxy servlet is really easy to use and has the features we need and it works with HTTPS, so thank you for that!

**Comment by dsmiley** [comment]
I will attempt to get this into the proxy servlet soonish; I'm on vacation this week though.  I'll try to put this into a separate HttpClient subclass.  I won't use commons-beanutils; it's just using reflection -- something that can be done manually easily enough.

**Comment by pvogt-svc** [comment]
Thanks a lot. In our business "soonish" is measured in weeks / months, so no pressure from our side :-)

Yes, it's only reflection. I'm still thinking that there should be a better / more regular way to do this. The example in the JSSE guide from Oracle does not use reflection. The question is, if there is a comparably easy way to do this with apache httpclient.

**Comment by dsmiley** [comment]
Please see the branch in which I incorporated this feature.  Calling setHost via reflection was super-easy without commons-beanutils; no regrets there :-).  Two things concern me:
* This isn't tested.  Can you offer suggestions?  Could Jetty be used in a test to respond over SSL and furthermore would there be a way of knowing what hostname the client sent?
* HttpClient 4.1 & 4.2 *should* still be supported by this proxy but the approach I have been taken thus far to verify it won't work any more.  I use HttpClient 4.3 classes now.  I'd like to be able to compile with one version and test with another.  Any ideas on how to do that?

**Comment by pvogt-svc** [comment]
Thx for the branch with the implementation. I can't really say whether the implementation is ok, because I do not have enough knowledge about apache httpclient with SSL. Perhaps a little bit of help about the problems with SSL and SNI may be the paper [Java’s SSLSocket: How Bad APIs Compromise Security by Georg Lukas](http://www.sicherheitsforschung-magdeburg.de/uploads/journal/MJS_034_Lukas_Java.pdf) chapter 1.2.2, that I read yesterday.

ad testing: sorry, I can't really help there as well. I don't use jetty that often and I have no idea how to test SSL with it. According to the [Jetty Documentation SSL / SNI](https://eclipse.org/jetty/documentation/current/configuring-ssl.html#d0e4828) you have to generate a keystore with the certificates for SSL / SNI with jetty. But how to test different hostnames when connecting to "localhost"? Perhaps it is possible to connect to jetty with SSL by IP and send a hostname for which a certificate is provided in the keystore.

ad HttpClient 4.1 & 4.2: For testing: perhaps using maven profiles and setting the `httpclient.test.version` to a different version? I have not seen something like the `<classifier>` in maven before, but it may help here?
